### PR TITLE
#60: multiple file selection

### LINKED ISSUE #60: "Open in explorer" button is not working when the file doesn't exist on disk
The "open in explorer" button usually opens the windows explorer in the file's folder *and* selects the file.
If the file doesn't exist, the explorer is opened at a completely different folder. This should be changed so that if the file doesn't exist, then the deepest folder that exists should be opened:

If the file path is C:\foo\bar\fizz.txt:
- If fizz.txt exists, then C:\foo\bar should be opened and fizz.txt should be selected
- If fizz.txt doesn't exist, then C:\foo\bar should be opened
- If bar doesn't exist, then C:\foo should be opened
- etc...

### PR DESCRIPTION
**Pull request by oleg-abr**
Fixes #60 

Changes proposed in this pull request:
Open explorer with multiple files selected (merged data source)

Any expected problems concerning backwards compatibility of existing plugins?  
No

Any expected problems concerning backwards compatibility of existing user settings?  
No

Does this break existing user workflows?  
No

### CODE REVIEW

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 1]
```diff
@@ -383,7 +383,7 @@ public void TestReadAll1()
 
 				file.Dispose();
 
-				file.GetValue(LogFileProperties.StartTimestamp).Should().Be(new DateTime(2015, 10, 7, 19, 50, 58, 982));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 2]
```diff
@@ -355,7 +355,7 @@ public void TestRead2LogEntries()
 						new LogFileSection(5, 1)
 					});
 
-				file.GetValue(LogFileProperties.StartTimestamp).Should().Be(new DateTime(2015, 10, 7, 19, 50, 58, 982));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 3]
```diff
@@ -289,12 +289,12 @@ public void TestOpenBeforeCreate()
 				new Action(() => logFile = new TextLogFile(_scheduler, fileName)).ShouldNotThrow();
 
 				logFile.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue(TimeSpan.FromSeconds(5));
-				logFile.Property(x => x.GetValue(LogFileProperties.EmptyReason)).ShouldEventually().Be(ErrorFlags.SourceDoesNotExist, TimeSpan.FromSeconds(5),
+				logFile.Property(x => x.Error).ShouldEventually().Be(ErrorFlags.SourceDoesNotExist, TimeSpan.FromSeconds(5),
 				                                                           "Because the specified file doesn't exist");
 
 				File.WriteAllText(fileName, "Hello World!");
 
-				logFile.Property(x => x.GetValue(LogFileProperties.EmptyReason)).ShouldEventually().Be(ErrorFlags.None, TimeSpan.FromSeconds(5),
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 4]
```diff
@@ -289,12 +289,12 @@ public void TestOpenBeforeCreate()
 				new Action(() => logFile = new TextLogFile(_scheduler, fileName)).ShouldNotThrow();
 
 				logFile.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue(TimeSpan.FromSeconds(5));
-				logFile.Property(x => x.GetValue(LogFileProperties.EmptyReason)).ShouldEventually().Be(ErrorFlags.SourceDoesNotExist, TimeSpan.FromSeconds(5),
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 5]
```diff
@@ -186,7 +186,7 @@ public void TestGetSection1()
 			{
 				file.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue(TimeSpan.FromSeconds(20), "because we should be able to read the entire file in a few seconds");
 				file.Count.Should().Be(165342);
-				file.GetValue(LogFileProperties.StartTimestamp).Should().Be(new DateTime(2015, 10, 7, 19, 50, 58, 982));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 6]
```diff
@@ -168,9 +168,9 @@ public void TestExists()
 				new Action(() => logFile = new TextLogFile(_scheduler, File2Lines)).ShouldNotThrow();
 
 				logFile.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue(TimeSpan.FromSeconds(5));
-				logFile.Property(x => x.GetValue(LogFileProperties.EmptyReason)).ShouldEventually().Be(ErrorFlags.None, TimeSpan.FromSeconds(5));
+				logFile.Property(x => x.Error).ShouldEventually().Be(ErrorFlags.None, TimeSpan.FromSeconds(5));
 
-				logFile.GetValue(LogFileProperties.EmptyReason).Should().Be(ErrorFlags.None, "Because the specified file does exist");
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 7]
```diff
@@ -168,9 +168,9 @@ public void TestExists()
 				new Action(() => logFile = new TextLogFile(_scheduler, File2Lines)).ShouldNotThrow();
 
 				logFile.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue(TimeSpan.FromSeconds(5));
-				logFile.Property(x => x.GetValue(LogFileProperties.EmptyReason)).ShouldEventually().Be(ErrorFlags.None, TimeSpan.FromSeconds(5));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 8]
```diff
@@ -148,9 +148,9 @@ public void TestDoesNotExist()
 				new Action(() => logFile = new TextLogFile(_scheduler, "dadwdawdw")).ShouldNotThrow();
 
 				logFile.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue(TimeSpan.FromSeconds(5));
-				logFile.Property(x => x.GetValue(LogFileProperties.EmptyReason)).ShouldEventually().Be(ErrorFlags.SourceDoesNotExist, TimeSpan.FromSeconds(5));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs [hunk 9]
```diff
@@ -148,9 +148,9 @@ public void TestDoesNotExist()
...
+				logFile.Property(x => x.Error).ShouldEventually().Be(ErrorFlags.SourceDoesNotExist, TimeSpan.FromSeconds(5));
 
-				logFile.GetValue(LogFileProperties.EmptyReason).Should().Be(ErrorFlags.SourceDoesNotExist, "Because the specified file doesn't exist");
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileAcceptanceTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs [hunk 10]
```diff
@@ -98,8 +98,8 @@ public void TestLive1And2()
 			{
 				merged.Property(x => x.Count).ShouldEventually().Be(19, TimeSpan.FromSeconds(5),
 				                                                    "Because the merged file should've been finished");
-				merged.Property(x => x.GetValue(LogFileProperties.Size)).ShouldEventually().Be(source0.GetValue(LogFileProperties.Size) + source1.GetValue(LogFileProperties.Size));
-				merged.Property(x => x.GetValue(LogFileProperties.StartTimestamp)).ShouldEventually().Be(source0.GetValue(LogFileProperties.StartTimestamp));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs [hunk 11]
```diff
@@ -98,8 +98,8 @@ public void TestLive1And2()
 			{
 				merged.Property(x => x.Count).ShouldEventually().Be(19, TimeSpan.FromSeconds(5),
 				                                                    "Because the merged file should've been finished");
-				merged.Property(x => x.GetValue(LogFileProperties.Size)).ShouldEventually().Be(source0.GetValue(LogFileProperties.Size) + source1.GetValue(LogFileProperties.Size));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs [hunk 12]
```diff
@@ -71,8 +71,8 @@ public void Test2SmallSources()
 
 				merged.Property(x => x.Count).ShouldEventually().Be(8, TimeSpan.FromSeconds(5),
 																	"Because the merged file should've been finished");
-				merged.Property(x => x.GetValue(LogFileProperties.Size)).ShouldEventually().Be(source0.GetValue(LogFileProperties.Size) + source1.GetValue(LogFileProperties.Size));
-				merged.Property(x => x.GetValue(LogFileProperties.StartTimestamp)).ShouldEventually().Be(source1.GetValue(LogFileProperties.StartTimestamp));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs [hunk 13]
```diff
@@ -71,8 +71,8 @@ public void Test2SmallSources()
 
 				merged.Property(x => x.Count).ShouldEventually().Be(8, TimeSpan.FromSeconds(5),
 																	"Because the merged file should've been finished");
-				merged.Property(x => x.GetValue(LogFileProperties.Size)).ShouldEventually().Be(source0.GetValue(LogFileProperties.Size) + source1.GetValue(LogFileProperties.Size));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs [hunk 14]
```diff
@@ -40,8 +40,8 @@ public void Test20Mb()
 				merged.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue();
 
 				merged.Count.Should().Be(source.Count);
-				merged.GetValue(LogFileProperties.Size).Should().Be(source.GetValue(LogFileProperties.Size));
-				merged.GetValue(LogFileProperties.StartTimestamp).Should().Be(source.GetValue(LogFileProperties.StartTimestamp));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs [hunk 15]
```diff
@@ -40,8 +40,8 @@ public void Test20Mb()
 				merged.Property(x => x.EndOfSourceReached).ShouldEventually().BeTrue();
 
 				merged.Count.Should().Be(source.Count);
-				merged.GetValue(LogFileProperties.Size).Should().Be(source.GetValue(LogFileProperties.Size));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/MergedLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/FilteredLogFileTest.cs [hunk 16]
```diff
@@ -41,7 +41,7 @@ public void TestFilter1()
 				using (FilteredLogFile filtered = file.AsFiltered(_scheduler, null, Filter.Create("info")))
 				{
 					filtered.Property(x => x.Count).ShouldEventually().Be(5, TimeSpan.FromSeconds(5));
-					filtered.GetValue(LogFileProperties.StartTimestamp).Should().Be(new DateTime(2015, 10, 7, 19, 50, 58, 982));
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/FilteredLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 17]
```diff
@@ -65,8 +63,8 @@ public void TestCtor()
 			var info = new FileInfo(_fname);
 			_scheduler.RunOnce();
 
-			_file.GetValue(LogFileProperties.LastModified).Should().Be(info.LastWriteTime);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 18]
```diff
@@ -65,8 +63,8 @@ public void TestCtor()
...
-			_file.GetValue(LogFileProperties.Created).Should().Be(info.CreationTime);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 19]
```diff
@@ -126,16 +124,16 @@ public void TestDoesNotExist()
 		{
 			_file = new TextLogFile(_scheduler, _fname);
 			_scheduler.RunOnce();
-			_file.GetValue(LogFileProperties.EmptyReason).Should().Be(ErrorFlags.None);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 20]
```diff
@@ -126,16 +124,16 @@ public void TestDoesNotExist()
...
-			_file.GetValue(LogFileProperties.Created).Should().NotBe(DateTime.MinValue);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 21]
```diff
@@ -126,16 +124,16 @@ public void TestDoesNotExist()
...
+			_file.Error.Should().Be(ErrorFlags.None);
+			_file.Created.Should().NotBe(DateTime.MinValue);
 
 			_streamWriter?.Dispose();
 			_stream?.Dispose();
 			File.Delete(_fname);
 			_scheduler.RunOnce();
 
-			_file.GetValue(LogFileProperties.EmptyReason).Should().Be(ErrorFlags.SourceDoesNotExist);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 22]
```diff
@@ -126,16 +124,16 @@ public void TestDoesNotExist()
...
-			_file.GetValue(LogFileProperties.Created).Should().BeNull();
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 23]
```diff
@@ -149,9 +147,9 @@ public void TestExclusiveAccess()
 				_file = new TextLogFile(_scheduler, _fname);
 				_scheduler.RunOnce();
 
-				_file.GetValue(LogFileProperties.EmptyReason).Should().Be(ErrorFlags.SourceCannotBeAccessed);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 24]
```diff
@@ -149,9 +147,9 @@ public void TestExclusiveAccess()
...
-				_file.GetValue(LogFileProperties.Created).Should().NotBe(DateTime.MinValue);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 25]
```diff
@@ -149,9 +147,9 @@ public void TestExclusiveAccess()
...
-				_file.GetValue(LogFileProperties.Created).Should().Be(new FileInfo(_fname).CreationTime);
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: This merge doesn't look right, GetValue(...) is the new way to retrieve data, please don't revert that.

File: Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs [hunk 26]
```diff
@@ -361,30 +359,5 @@ public void TestGetColumnDeltaTime2()
 				null
 			});
 		}
-
```
Comment (Kittyfisto) [inline Tailviewer.AcceptanceTests/BusinessLogic/LogFiles/TextLogFileTest.cs]: I don't see why the following 40 lines were removed.

**Comment by oleg-abr** [comment]
Looks like merge went wrong

**Pull request closed without merging**